esp/esp/program/models/__init__.py

```diff
@@ -536,26 +536,121 @@ def volunteers_union(self, QObject = False):
         else:
             return ESPUser.objects.filter(union).distinct()
 
+    # Don't bother caching this.  Everything it calls is cached to the extent
+    # possible, with different dependencies.
+    def user_can_join(self, user):
+        """Can this user join this program?
+
+        The program cap may be set by either making Program.program_size_max
+        nonzero or by the tag "program_size_by_grade" (which should be a json
+        object of grade -> cap).  The latter overrides the former.  If there is
+        a (possibly per-grade) cap, and the program is at or above that cap, a
+        user can only join a program if they are already registered for it, or
+        if they have the OverrideFull permission.
+
+        The tag's value, if set, should be a json object.  The keys should be
+        grades (as strings, e.g. "9") or inclusive ranges of grades (separated
+        by hyphens, e.g.  "7-9"), and the values should be caps for that range.
+        Each cap will be checked, so having overlapping caps is possible,
+        although it may hurt performance.
+        """
+        # TODO(benkraft): maybe this shouldn't be a Tag.  For now it basically
+        # has to be because I don't want to deal with migrations until after
+        # we're on django 1.8.
```

Comment:
The other problem is that it's more than just a number -- currently as JSON.  I'm not a huge fan of storing JSON -- we have a DB for structured data, so let's use it.  But making a model for this seems kind of overkill.